File: backend/src/photonflow/core/signal.py
"""Signal data structure."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, Optional

import torch


@dataclass
class Signal:
    data: torch.Tensor
    fs: float
    t0: float = 0.0
    center_freq: Optional[float] = None
    pol_mode: str = "scalar"
    meta: Dict[str, Any] = field(default_factory=dict)

    def clone(self) -> "Signal":
        return Signal(
            data=self.data.clone(),
            fs=self.fs,
            t0=self.t0,
            center_freq=self.center_freq,
            pol_mode=self.pol_mode,
            meta=dict(self.meta),
        )

    def time(self) -> torch.Tensor:
        n = self.data.shape[-1]
        device = self.data.device
        if self.data.dtype in (torch.complex64, torch.complex128):
            dtype = torch.float32 if self.data.dtype == torch.complex64 else torch.float64
        else:
            dtype = self.data.dtype
        return self.t0 + torch.arange(n, device=device, dtype=dtype) / self.fs

    def is_optical(self) -> bool:
        return self.center_freq is not None

    def is_jones(self) -> bool:
        return self.pol_mode == "jones"

File: photonflow-cpp/src/core/signal.cpp
/**
 * @file signal.cpp
 * @brief Signal class implementation.
 */

#include "photonflow/core/signal.hpp"

namespace photonflow {

Signal::Signal(const Eigen::VectorXcd& data, double fs, double t0)
    : data(data)
    , fs(fs)
    , t0(t0)
    , center_freq(std::nullopt)
    , pol_mode("scalar")
{}

Signal Signal::clone() const {
    Signal copy(data, fs, t0);
    copy.center_freq = center_freq;
    copy.pol_mode = pol_mode;
    copy.meta = meta;
    return copy;
}

Eigen::VectorXd Signal::time() const {
    const Eigen::Index n = data.size();
    Eigen::VectorXd t(n);
    for (Eigen::Index i = 0; i < n; ++i) {
        t[i] = t0 + static_cast<double>(i) / fs;
    }
    return t;
}

bool Signal::is_optical() const {
    return center_freq.has_value();
}

bool Signal::is_jones() const {
    return pol_mode == "jones";
}

} // namespace photonflow
